crates/app-host/src/event_log_fetcher.rs

```diff
@@ -0,0 +1,210 @@
+use alloy::{eips::BlockNumberOrTag, primitives::Address, sol_types::SolEvent};
+use base::eth::EthError;
+use crate::{l1_client::L1Client, types::{CommitBatchEvent, FinalizeBatchEvent, ScrollChain}};
+use std::time::Duration;
+use tokio::{sync::mpsc::Sender, time::interval};
+use anyhow::Result;
+use event_log_parser::EventLogParser;
+
+base::stack_error! {
+    #[derive(Debug)]
+    name: EventLogError,
+    stack_name: EventLogErrorStack,
+    error: {
+        Eth(EthError),
+        Fetcher(std::borrow::Cow<'static, str>),
+        Parser(std::borrow::Cow<'static, str>),
+        Other(std::borrow::Cow<'static, str>),
+    },
+    wrap: {
+    },
+    stack: {}
+}
+
+impl From<EthError> for EventLogError {
+    fn from(value: EthError) -> Self {
+        Self::Eth(value)
+    }
+}
+
+pub struct EventLogFetcher {
+    event_log_parser: EventLogParser,
+    l1_client: L1Client,
+    scroll_chain_address: Address,
+    max_size_per_fetch: u64,
+    fetched_block_number: u64,
+    finalized_block_number: u64,
+
+    commit_batch_tx: Sender<CommitBatchEvent>,
+    finalize_batch_tx: Sender<FinalizeBatchEvent>,
+}
+
+impl EventLogFetcher {
+    pub fn new(
+        commit_batch_tx: Sender<CommitBatchEvent>,
+        finalize_batch_tx: Sender<FinalizeBatchEvent>,) -> Self {
+        todo!()
+    }
+
+    async fn get_latest_finalized_block(&self) -> Result<u64, EventLogError> {
+        let block = self.l1_client.get_block_by_number(BlockNumberOrTag::Finalized).await.map_err(EthError::from)?;
+        if block.is_none() {
+            return Err(EventLogError::Fetcher("get empty block".into()));
+        }
+        match block.unwrap().header.number {
+            Some(n) => Ok(n),
+            None => Err(EventLogError::Fetcher("no block number in header".into()))
+        }
+    }
+
+    async fn fetch_logs(&self) -> Result<(), EventLogError> {
+        let last_finalize_block = self.get_latest_finalized_block().await?;
```

Comment:
Why only fetch up to latest finalized block? To avoid reorg handling?

Reply:
yep, it's the same as the one in our go-ethereum's l1_event fetch logic

Reply:
I don't think this will be suitable here, the 12-15min delay will be too large. We want to submit SGX proof shortly after the ZK proof was submitted.